allennlp/nlp_api.py

```diff
@@ -0,0 +1,289 @@
+        in your input, like "passage_tokens", or "question_tokens".  Because our
+        :class:`~allennlp.data.fields.TextField` returns a `list` of arrays, one for each
+        :class:`~allennlp.data.token_indexers.TokenIndexer`, you should typically specify a `list`
+        of ``Modules`` here, one to process each input array and produce a vector for each token.
+        E.g., if you convert passage words into two different word ids, plus a sequence of
+        character ids, you should have three modules for "passage_tokens" to do three separate
+        embeddings given the different arrays (one of which could actually be a CNN encoder).  The
+        three separate arrays then get concatenated and returned.
+
+        Alternatively, instead of giving a list of ``Modules``, you could give a single ``Module``
+        that knows how to process the whole list of inputs, returning a single ``Tensor``.
+
+        This abstraction is key to allowing you to easily switch between word embeddings and
+        character-level encoders, or token-in-context embeddings produced by some pre-trained
+        model, or any combination of these.  Instead of instantiating your own embedding layers, or
+        character-level CNNs, you should call :func:`embed_text_field` to do that for you, so it is
+        easy to try new representations later.  See the documentation of that function for a little
+        more detail.
+    recurrent_layers : ``Dict[str, torch.nn.Module]``, optional (default=``{}``)
+        A mapping from names to recurrent layers, like LSTMs and GRUs.  The name here corresponds
+        to some part of your `model`, like "passage_encoder", not the `type` of the recurrent
+        layer.
+    default_recurrent_layer : ``Callable[[], torch.nn.Module]``, optional (default=``None``)
+        A function that returns a new recurrent layer.  Giving this makes it so you can use unique
+        names for all recurrent layers in your model, but still only have to specify one layer
+        type.  See :func:`get_recurrent_layer` for more detail.
+    recurrent_cells : Dict[str, torch.nn.Module], optional (default=``{}``)
+        A mapping from names to recurrent cells, for building, e.g., a TreeLSTM.  Similarly to
+        ``recurrent_layers``, the names correspond to places in your model where these cells get
+        used, not to anything about what kind of cell you're using.
+    default_recurrent_cell : ``Callable[[], torch.nn.Module]``, optional (default=``None``)
+        A function that returns a new recurrent cell.  Giving this makes it so you can use unique
+        names for all recurrent cells in your model, but still only have to specify one cell type.
+        See :func:`get_recurrent_cell` for more detail.
+    """
+    def __init__(self,
+                 embedders: Dict[str, Union[torch.nn.Module, List[torch.nn.Module]]] = None,
+                 recurrent_layers: Dict[str, torch.nn.Module] = None,
```

Comment:
That's a good point.  I started out thinking I'd do `encoders` and `seq2seq_encoders`, as we currently have, but then realized that pytorch just always does `return_sequences=True`, so I thought we wouldn't need to make a distinction between the two.  I forgot about CNNs and TreeLSTMs and such.  I'll switch it back to `encoders` and `seq2seq_encoders` - it just means we'll need some thin wrappers around pytorch's RNNs, I think.  So, when you're looking through this, imagine there's `get_encoder` and `get_seq2seq_encoder`, instead of `get_recurrent_layer`, both of which have basically identical behavior to this.